File: backend/api/routers/job_runs.py
import logging

from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import Depends, Response, status, APIRouter

from api.dependencies import get_session, get_transactional_session
from db.enums import CameraType
from db.models.job_run import JobRun
from schemas.job_run import CreateJobRunRequest, JobRunResponse
from services.errors import JobRunNotFoundError
from services.job_run_service import JobRunService
from utilities.service_builder_utilities import build_job_run_service

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s"
)
logger = logging.getLogger(__name__)

job_runs_router = APIRouter(
    prefix="/job-runs",
)

@job_runs_router.get("/", response_model=list[JobRunResponse])
async def list_job_runs(
    session: AsyncSession = Depends(get_session),
) -> list[JobRun]:
    logging.info("get job runs")
    job_run_service = build_job_run_service(session=session)
    return await job_run_service.list_job_runs()


@job_runs_router.get("/{route_id:path}/{job_def_id}/{job_run_num}/{camera}", response_model=JobRunResponse)
async def get_job_run(
    job_def_id: int,
    job_run_num: int,
    route_id: str,
    camera: CameraType,
    session: AsyncSession = Depends(get_session),
) -> JobRun:
    logging.info(f"Getting job run ({job_def_id=}, {job_run_num=}, {route_id=})")
    job_run_service = build_job_run_service(session=session)
    job_run = await job_run_service.get_job_run(
        job_def_id=job_def_id,
        job_run_num=job_run_num,
        route_id=route_id,
        camera=camera,
    )
    if job_run is None:
        raise JobRunNotFoundError(
            job_run_num=job_run_num,
            job_def_id=job_def_id,
            route_id=route_id
        )
    return job_run


@job_runs_router.post("/", response_model=JobRunResponse, status_code=status.HTTP_201_CREATED)
async def create_job_run(
    payload: CreateJobRunRequest,
    session: AsyncSession = Depends(get_transactional_session),
) -> JobRun:
    logging.info("create job run")
    job_run_service: JobRunService = build_job_run_service(session=session)
    return await job_run_service.create_job_run(
        job_def_id=payload.job_def_id,
        route_id=payload.route_id,
        camera=payload.camera,
    )


@job_runs_router.delete("/")
async def delete_job_run(
    job_def_id: int,
    job_run_num: int,
    route_id: str,
    camera: CameraType,
    session: AsyncSession = Depends(get_transactional_session),
):
    logging.info("delete job run")
    job_run_service = build_job_run_service(session=session)
    await job_run_service.delete_job_run(
        job_run_num=job_run_num,
        job_def_id=job_def_id,
        route_id=route_id,
        camera=camera
    )
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/job_runs.ts
import { apiFetch } from "./client";
import { mapJobRun } from "./types";
import type { JobRun, JobRunResponse } from "./types";

export async function listJobRuns(): Promise<JobRun[]> {
  const data = await apiFetch<JobRunResponse[]>("/job-runs/");
  return data.map(mapJobRun);
}

export async function listJobRunsForRoute(routeId: string): Promise<JobRun[]> {
  const data = await apiFetch<JobRunResponse[]>(`/job-runs/by-route/${routeId}`);
  return data.map(mapJobRun);
}

export async function createJobRun(params: {
  jobDefId: number;
  routeId: string;
  camera: string;
}): Promise<JobRun> {
  const data = await apiFetch<JobRunResponse>("/job-runs/", {
    method: "POST",
    body: JSON.stringify({
      job_def_id: params.jobDefId,
      route_id: params.routeId,
      camera: params.camera,
    }),
  });
  return mapJobRun(data);
}

export async function deleteJobRun(params: {
  jobDefId: number;
  jobRunNum: number;
  routeId: string;
  camera: string;
}): Promise<void> {
  const qs = new URLSearchParams({
    job_def_id: String(params.jobDefId),
    job_run_num: String(params.jobRunNum),
    route_id: params.routeId,
    camera: params.camera,
  });
  await apiFetch<void>(`/job-runs/?${qs}`, { method: "DELETE" });
}

export async function cancelJobRun(params: {
  jobDefId: number;
  jobRunNum: number;
  routeId: string;
  camera: string;
}): Promise<JobRun> {
  const qs = new URLSearchParams({
    job_def_id: String(params.jobDefId),
    job_run_num: String(params.jobRunNum),
    route_id: params.routeId,
    camera: params.camera,
  });
  const data = await apiFetch<JobRunResponse>(`/job-runs/cancel?${qs}`, { method: "POST" });
  return mapJobRun(data);
}

export async function requeueJobRun(params: {
  jobDefId: number;
  jobRunNum: number;
  routeId: string;
  camera: string;
}): Promise<JobRun> {
  const qs = new URLSearchParams({
    job_def_id: String(params.jobDefId),
    job_run_num: String(params.jobRunNum),
    route_id: params.routeId,
    camera: params.camera,
  });
  const data = await apiFetch<JobRunResponse>(`/job-runs/requeue?${qs}`, { method: "POST" });
  return mapJobRun(data);
}
